<launch>
  <arg name="robot_namespace" default="husky"/>
  <arg name="base_namespace" default="base1"/>

  <arg name="b_use_hero" default="false"/>

  <!-- List of robots with hovermap mounted -->
  <arg name="hovermap_robots" value="spot3 spot4 husky1 husky2 husky3" />
  <arg name="b_use_hovermap" value="$(eval robot_namespace in hovermap_robots)" />

  <!-- Select odometry source -->
  <arg if="$(eval b_use_hero)" name="odom_input" value="hero/odometry"/>
  <arg if="$(eval not b_use_hero and b_use_hovermap)" name="odom_input" value="hvm/odometry"/>
  <arg if="$(eval not b_use_hero and not b_use_hovermap)" name="odom_input" value="locus/odometry"/>

  <!-- optional filename of the zip archive of a pose graph to be loaded -->
  <arg name="load_pose_graph_file" default="" />


  <!-- The robot prefix dependent of robot_namespace -->
  <arg name="robot_prefix" default="a" if="$(eval robot_namespace == 'husky1')" />
  <arg name="robot_prefix" default="b" if="$(eval robot_namespace == 'husky2')"/>
  <arg name="robot_prefix" default="c" if="$(eval robot_namespace == 'husky3')"/>
  <arg name="robot_prefix" default="d" if="$(eval robot_namespace == 'husky4')"/>
  <arg name="robot_prefix" default="e" if="$(eval robot_namespace == 'spot1')"/>
  <arg name="robot_prefix" default="f" if="$(eval robot_namespace == 'spot2')"/>
  <arg name="robot_prefix" default="g" if="$(eval robot_namespace == 'spot3')"/>
  <arg name="robot_prefix" default="h" if="$(eval robot_namespace == 'spot4')"/>
  <arg name="robot_prefix" default="i" if="$(eval robot_namespace == 'aquila1')"/>
  <arg name="robot_prefix" default="j" if="$(eval robot_namespace == 'aquila2')"/>
  <arg name="robot_prefix" default="k" if="$(eval robot_namespace == 'kaist1')"/>
  <arg name="robot_prefix" default="l" if="$(eval robot_namespace == 'zoe1')"/>
  <arg name="robot_prefix" default="m" if="$(eval robot_namespace == 'zoe1')"/>
  <arg name="robot_prefix" default="x" if="$(eval robot_namespace == 'xmaxx1')"/>

  <!-- The artifact prefix dependent of robot_namespace -->
  <arg name="artifact_prefix" default="A" if="$(eval robot_namespace == 'husky1')" />
  <arg name="artifact_prefix" default="B" if="$(eval robot_namespace == 'husky2')"/>
  <arg name="artifact_prefix" default="C" if="$(eval robot_namespace == 'husky3')"/>
  <arg name="artifact_prefix" default="D" if="$(eval robot_namespace == 'husky4')"/>
  <arg name="artifact_prefix" default="E" if="$(eval robot_namespace == 'spot1')"/>
  <arg name="artifact_prefix" default="F" if="$(eval robot_namespace == 'spot2')"/>
  <arg name="artifact_prefix" default="G" if="$(eval robot_namespace == 'spot3')"/>
  <arg name="artifact_prefix" default="H" if="$(eval robot_namespace == 'spot4')"/>
  <arg name="artifact_prefix" default="J" if="$(eval robot_namespace == 'aquila2')"/>
  <arg name="artifact_prefix" default="K" if="$(eval robot_namespace == 'kaist1')"/>
  <arg name="artifact_prefix" default="L" if="$(eval robot_namespace == 'zoe1')"/>
  <arg name="artifact_prefix" default="M" if="$(eval robot_namespace == 'zoe1')"/>
  <arg name="artifact_prefix" default="X" if="$(eval robot_namespace == 'xmaxx1')"/>

  <group ns="$(arg robot_namespace)">


    <!-- SLAM -->
    <node pkg="lamp"
          name="lamp"
          type="lamp_node"
          output="screen">
          <!-- launch-prefix="xterm -e gdb - -args"> -->

      <!-- Load graph -->
      <param name="load_graph" value="$(arg load_pose_graph_file)"/>

      <!-- Use fixed covariances, rather than computed -->
      <param if="$(eval b_use_hovermap)" name="b_use_fixed_covariances" value="true" />
      <param unless="$(eval b_use_hovermap)" name="b_use_fixed_covariances" value="false" />

      <!-- Topics -->
      <remap from="~pcld" to="body_filter/output" unless="$(arg b_use_hovermap)"/>
      <remap from="~pcld" to="hvm/body_filter/output" if="$(arg b_use_hovermap)"/>
      <!-- <remap from="~pcld" to="velodyne_points/transformed"/> -->
      <!-- <remap from="~pcld" to="lo_frontend/local_scan_map" /> -->
      <!-- Note that the transformed points come from the transforming node launched with lo_frontend -->
      <!-- <remap from="~lio_odom" to="lo_frontend/odometry"/> -->
      <!-- <remap from="~lio_odom" to="hero/lion/odom"/> -->
      <remap from="~lio_odom" to="$(arg odom_input)"/>
      <!-- <remap from="~lio_odom" to="$(arg odom_input_husky)" unless="$(eval robot_namespace.startswith('spot'))"/> -->
      <!-- <remap from="~lio_odom" to="spot_frontend/odometry" if="$(eval robot_namespace.startswith('spot'))"/> -->
      <remap from="~vio_odom" to="visual_inertial_odometry_topic_currently_not_used"/>
      <remap from="~wio_odom" to="wheel_inertial_odometry_topic_currently_not_used"/>
      <remap from="~optimized_values" to="lamp_pgo/optimized_values"/>

      <remap from="~artifact" to="~artifact_global" />
      <remap from="~artifact_relative" to="artifact/update" />

      <remap from="~imu_topic" to="vn100/imu_wori_wcov"/>
      <remap from="~stationary_topic" to="stationary_accel"/>

      <!-- LAMP settings -->
      <rosparam file="$(find lamp)/config/lamp_settings.yaml" subst_value="true"/>
      <rosparam file="$(find lamp)/config/lamp_init_noise.yaml" subst_value="true"/>
      <rosparam file="$(find lamp)/config/precision_parameters.yaml" subst_value="true"/>

      <!-- Factor handlers settings -->
      <rosparam file="$(find factor_handlers)/config/odom_parameters.yaml" subst_value="true"/>
      <rosparam file="$(find factor_handlers)/config/imu_parameters.yaml" subst_value="true"/>
      <rosparam file="$(find factor_handlers)/config/april_parameters.yaml" subst_value="true"/>

      <!-- Rates -->
      <rosparam file="$(find lamp)/config/lamp_rates.yaml"/>

      <!-- Frames -->
      <rosparam file="$(find lamp)/config/lamp_frames.yaml" subst_value="true"/>

      <!-- Point cloud filter -->
      <rosparam file="$(find lamp)/config/filter_parameters.yaml"/>

      <!-- Point cloud mapper -->
      <rosparam file="$(find point_cloud_mapper)/config/parameters.yaml"/>

      <param name="robot_prefix" value="$(arg robot_prefix)" />
    </node>


    <!-- Pose graph optimizer -->
    <!-- <node pkg="lamp_pgo"
          name="lamp_pgo"
          type="lamp_pgo_node"
          output="screen">
      <remap from="~pose_graph_to_optimize" to="lamp/pose_graph_to_optimize" />
      <rosparam file="$(find lamp_pgo)/config/pgo_parameters.yaml" subst_value="true"/>
    </node> -->

    <!-- Loop Closure  -->
    <!-- <include file="$(find loop_closure)/launch/loop_closure_modules.launch"/> -->

  </group>

  <!-- Base Station graph merger  -->
  <include file="$(find pose_graph_merger)/launch/two_pose_graph_merge.launch">
    <arg name="robot_namespace" value="$(arg robot_namespace)"/>
    <arg name="base_namespace" value="$(arg base_namespace)"/>
  </include>

  <!-- Zero Velocity Detector  -->
  <include file="$(find localizer_zero_velocity_detector)/launch/very_stable_genius.launch">
    <arg name="robot_namespace" value="$(arg robot_namespace)"/>
  </include>

</launch>
